### PR TITLE
[BarPlot] - Have scale account for bar width

### LINKED ISSUE #1280: Scales in bar plots don't account for bar width
If a bar was placed on the edge of the scale, part of the bar could extend past the edge of the scale, causing that part of the bar to clip.  The scale should be extended to account for the bar.

### PR DESCRIPTION
**Pull request by bluong**
Update scale extents before render based on the width of the bar

[@bluong: 1/29 14:31] EDIT: Will only perform operation if width is manually set.  If width is auto calculated then #1544 should be used.

Fixes #1280

### CODE REVIEW

File: src/components/plots/barPlot.ts [hunk 1]
```diff
@@ -43,6 +43,65 @@ export module Plot {
...
+
+    public detach() {
+      this._datasetKeysInOrder.forEach((key) => {
+        var scaleKey = this.getID().toString() + "_" + key;
+        this._xScale._removeExtent(scaleKey, "bar-extent");
+      });
+      return this;
+    }
+
+    public remove() {
```
Comment (jtlan) [inline src/components/plots/barPlot.ts]: No `super()`?

File: src/components/plots/barPlot.ts [hunk 2]
```diff
@@ -43,6 +43,65 @@ export module Plot {
...
 
+    protected _onDatasetUpdate() {
+      super._onDatasetUpdate();
+      if (this._isAnchored) {
+        this._updateBarExtent();
+      }
+    }
+
+    public detach() {
+      this._datasetKeysInOrder.forEach((key) => {
```
Comment (jtlan) [inline src/components/plots/barPlot.ts]: `super()`?

**Comment by jtlan** [comment]
If the goal is to make sure the bars don't get drawn beyond the limits of the scale, wouldn't it make more sense to change the math in `_getBarPixelWidth()`?

**Comment by bluong** [comment]
I thought the goal was extending the scale to fit the bar width?

File: src/components/plots/barPlot.ts (hunk 1 above)
Comment (bluong) [inline src/components/plots/barPlot.ts]: github's being dumb as the super shhould be there?

**Comment by jtlan** [comment]
The primary goal is to have the bars not extend past the visible range, which could be done multiple ways.

**Comment by jtlan** [comment]
Unit test?

File: src/components/plots/barPlot.ts [hunk 3]
```diff
@@ -43,6 +43,67 @@ export module Plot {
...
+      return this;
+    }
+
+    public remove() {
+      super.remove();
+      this._datasetKeysInOrder.forEach((key) => {
+        var scaleKey = this.getID().toString() + "_" + key;
+        this._xScale._removeExtent(scaleKey, "bar-extent");
+      });
+      return this;
```
Comment (jtlan) [inline src/components/plots/barPlot.ts]: `remove()` calls `detach()` (see `AbstractComponent`), so you don't need to override this.

File: test/components/plots/barPlotTests.ts [hunk 4]
```diff
@@ -597,5 +597,26 @@ describe("Plots", () => {
...
+      plot.addDataset(simpleDataset)
+          .project("x", "a", xScale)
+          .project("y", "b", yScale)
+          .project("width", 200)
+          .renderTo(svg);
+
+      plot.getAllBars().each(function(selection) {
+        var bar = d3.select(this);
+        assert.operator(+bar.attr("x"), ">=", 0, "all rects right of left edge of scale");
+        assert.operator(+bar.attr("x") + +bar.attr("width"), "<=", 500, "all rects left of right edge of scale");
```
Comment (jtlan) [inline test/components/plots/barPlotTests.ts]: maybe use `scale.range()` to avoid the numbers?

File: src/components/plots/barPlot.ts [hunk 5]
```diff
@@ -43,6 +43,63 @@ export module Plot {
...
+      if (this._isAnchored) {
+        this._updateBarExtent();
+      }
+    }
+
+    public detach() {
+      super.detach();
+      this._datasetKeysInOrder.forEach((key) => {
+        var scaleKey = this.getID().toString() + "_" + key;
+        this._xScale._removeExtent(this.getID().toString(), "bar-extent");
```
Comment (jtlan) [inline src/components/plots/barPlot.ts]: I wonder if `getID()` should just always return strings. There's no point in doing math on an ID...?

**Comment by jtlan** [comment]
[minor] Test could use `range()` instead of magic numbers.

`Plot.Bar` is getting pretty heavy. I wonder if there's some sort of streamlining refactor in our future...

**Comment by crmorford** [comment]
Can I get a hint on how to make this happen? I can't remember what usually causes this.

**Comment by crmorford** [comment]
this doesn't look resolved to me (look at far left/far right bars) http://jsfiddle.net/mkwvpw4x/2/

**Comment by crmorford** [comment]
Just talked offline, passing the fiddle: http://jsfiddle.net/mkwvpw4x/4/

**Comment by bluong** [comment]
Closing PR due to low priority and difficulties
